### PR TITLE
Allow configuration of test source inclusion and exclusion

### LINKED ISSUE #87: Allow specification of unit-test executable file name(s)
I noticed that the discovery of unit-tests scans every .exe file in the output directory. In our case, this is quite much and hence takes a lot of time: We have 70 .exe files in the output directory, of which only one includes the unit tests.

It would be great if it was possible to specificy which .exe file(s) should be scanned for unit tests.

Seems possible? In line with your other priorities and ideas?

### PR DESCRIPTION
**Pull request by netspiri**
Test sources can be filtered via inclusion and exclusion regular expression patterns.

Patterns are defined in the following manner:

```
<?xml version="1.0" encoding="utf-8"?>
<RunSettings>

    <!-- BoostTest Adapter -->
    <BoostTest>
        <Filters>
			<Exclude>
				<Pattern>BoostUnitTest.exe$</Pattern>
			</Exclude>
			<Include>
				<Pattern>NoExcludeTest.exe$</Pattern>
			</Include>
		</Filters>
    </BoostTest>

</RunSettings>
```

Users can omit the `Include` or `Exclude` sections to specify 'include-only' or 'include-everything-but' semantics.

Closes #87

### CODE REVIEW

File: BoostTestAdapter/BoostTestDiscoverer.cs [hunk 1]
```diff
@@ -102,6 +102,12 @@ public void DiscoverTests(IEnumerable<string> sources, IDiscoveryContext discove
 
             try
             {
+                // Filter out any sources which are not of interest
+                if ((settings.Filters != null) && (!settings.Filters.IsEmpty))
```
Comment (guwirth) [inline BoostTestAdapter/BoostTestDiscoverer.cs]: IsNullOrEmpty?

File: BoostTestAdapter/Settings/TestSourceFilter.cs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+                return false;
+            }
+
+            // Test if an include pattern matches the source file path
+            if (!IsNullOrEmpty(this.Include) && IsMatch(this.Include, source))
+            {
+                return true;
+            }
+
+            // If no exclude/include patterns are defined, accept any and all tests else accept tests if-and-only-if an exclude list is present without an include list
```
Comment (guwirth) [inline BoostTestAdapter/Settings/TestSourceFilter.cs]: Difficult to understand

File: BoostTestAdapter/BoostTestDiscoverer.cs (hunk 1 above)
Comment (netspiri) [inline BoostTestAdapter/BoostTestDiscoverer.cs]: Fixed in new push

**Comment by guwirth** [comment]
:+1:

**Pull request merged**